### DISCUSSION TITLE
Change how we create our release notes

### DISCUSSION THREAD

**Discussion by robertlaurin** [open-telemetry/opentelemetry-ruby#721] [Ideas]
While going through the release notes for our [`0.17.0`](https://github.com/open-telemetry/opentelemetry-ruby/pull/720) some [good observations](https://github.com/open-telemetry/opentelemetry-ruby/pull/720#issuecomment-824991873) were made about how how some of the breaking change notes were not exactly breaking.

For example [farday](https://github.com/open-telemetry/opentelemetry-ruby/pull/720/files#diff-6b8108d2c50d3daeb53032e989956edd30a1c181b60b57130b1030b5e75e2ba8R5) was marked as having a breaking change, but the PR that this message was derived from actually only changed the tests for Faraday.

Another example is a fairly [significant breaking](https://github.com/open-telemetry/opentelemetry-ruby/pull/720/files#r618560743) breaking change to the API was missing entirely.  

While the automated release note process has saved us a lot of effort along the way, I think the project is at a point where it makes sense for these release notes to be more closely curated by humans.  

I'm proposing we do away with the automated changelog generation and instead introduce something like an `unreleased` changelog section that is populated at the same time as the PRs that introduce the changes.  That way we can make sure that PR that changes multiple gems populates the appropriate notes for each gem, some gems may receive breaking changes, and some may not.

In addition to that I strongly believe our release notes should link back to the PRs that introduce the changes.

**Comment by fbogsany** [comment]
I think the automation is helpful as a starting point, but it would be nice if the suggested changelog additions happened earlier, in the PRs making the changes, rather than just in the release PR.

Perhaps some kind of bot would be suitable for this? Alternatively, a PR template with a checklist that included "update changelog".

**Reply by robertlaurin** [reply]
I do think a PR template would be good regardless so the expectation is clearly signalled to anyone contributing (or reminding us).  I agree about the automation, it would be nice if happened with each PR.  Not sure what the would look like exactly, a CI check of sorts, or something that a maintainer would trigger before merging the PR?

**Reply by robbkidd** [reply]
The changelog generation I've experienced before would use pull request titles for the one-liner entries in the log. That provided maintainers an opportunity to edit the entry as a part of review and prior to merging when the title and a `#nnnn` reference to the PR would become part of the generated CHANGELOG. Generally, PRs that were backwards breaking would include in their descriptions the ways in which users would adjust to the backwards-breaking behavior. At release time, a maintainer would collect those migration instructions to manually (😭 ) add to a RELEASE_NOTES for a particular version. 

Most of the changelog automation I know of is only capable of working with git commit summaries or PR titles, so there's no convenance of "and include these notes, too, please."

A PR template could help that human-in-the-loop migration instruction gathering. Maybe a section for:

<details><summary> "So, you've put a ! in your conventional commit. What do people need to do to adjust to this change?"</summary>

![Captain America sits on a chair and says, "So, you've got detention."](https://user-images.githubusercontent.com/517302/115767945-a8998080-a377-11eb-9b57-284310026842.gif)

</details>
